### PR TITLE
UI: Migrate Select from Popper.js tooltip to react-aria overlay

### PR DESCRIPTION
**Pull request by Sidnioulz**
> [!CAUTION]
> Must base onto next once the IconButton branch is merged! ⚠️⚠️⚠️ 

## What I did

This PR is part of the work done on WithTooltip replacement. It specifically addresses Select's use of WithTooltip.

Select's menu is not a tooltip, it's an interactive popover with a listbox role. We already have KB nav and aria markup working, all we were missing was:
* A popover placement logic that doesn't depend on Popper.js
* A hook to close the popover when clicking outside of its content

The PR uses low-level react-aria hooks to achieve this.

The miminalist popover inside `Select.tsx` will likely not be reusable. We have higher-level primitives for most use cases, but here I needed something very low-level to keep the existing KB nav logic in place.

#### The changes in this PR are covered in the following automated tests:

- [x] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

1. Run local instance
2. Go to http://localhost:6006/?path=/story/select-component--mouse-selection-multi
3. Play around with the stories

### Documentation

N/A

## Checklist for Maintainers

- [x] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

## Greptile Summary

Updated On: 2025-09-03 17:34:52 UTC

This PR migrates the Select component from using `WithTooltipPure` (based on Popper.js) to react-aria overlay hooks for better semantic correctness and accessibility. The Select component's dropdown menu was previously implemented as a tooltip, but semantically it should be an interactive popover with a listbox role.

The main changes include:

1. **New dependency**: Adds `react-aria-components@^1.12.1` to the core package to access overlay positioning and interaction hooks

2. **Custom MinimalistPopover component**: Replaces `WithTooltipPure` with a new component that uses three react-aria hooks:
   - `useOverlayPosition` for popover placement without Popper.js
   - `useInteractOutside` for handling clicks outside the popover
   - `useOverlay` for focus management and dismissal behavior

3. **Improved ARIA semantics**: Updates role assignments so single-select uses `button` role and multi-select uses `combobox` role, following proper accessibility guidelines

4. **Enhanced keyboard navigation**: Maintains existing keyboard shortcuts while adding Tab key handling to close the popover

5. **Visual consistency**: Adds `borderRadius: 2` to focused select options for consistent styling

The migration preserves all existing functionality including complex multi-select logic, keyboard navigation, and ARIA markup while moving to a more semantically correct and accessible implementation. The new MinimalistPopover is intentionally low-level and non-reusable to maintain the existing keyboard navigation behavior.

## Confidence score: 4/5

- This PR introduces significant architectural changes but follows established patterns and maintains existing functionality
- Score reflects the complexity of the migration and potential edge cases with the new react-aria dependency
- Pay close attention to the Select component implementation and test the popover positioning thoroughly

### CODE REVIEW

File: code/core/src/components/components/Select/SelectOption.tsx [hunk 1]
```diff
@@ -61,6 +61,7 @@ const Item = styled('li')(({ theme }) => ({
     background: theme.background.hoverable,
     outline: `2px solid ${theme.barSelectedColor}`,
     outlineOffset: 1,
+    borderRadius: 2,
```
Comment (ghengeveld) [inline code/core/src/components/components/Select/SelectOption.tsx:64]: We have a theme variable for border radius. Shouldn't we use that, or is this intentionally different?

File: code/core/src/components/components/Select/Select.tsx [hunk 2]
```diff
@@ -474,7 +458,85 @@ export const Select = forwardRef<HTMLButtonElement, SelectProps>(
...
+                  } else if (e.key === 'Home') {
+                    setActiveOption(options[0]);
+                  } else if (e.key === 'End') {
+                    setActiveOption(options[options.length - 1]);
+                  } else if (e.key === 'PageDown') {
+                    moveActiveOptionDown(PAGE_STEP_SIZE);
+                  } else if (e.key === 'PageUp') {
+                    moveActiveOptionUp(PAGE_STEP_SIZE);
+                  }
+                }}
```
Comment (ghengeveld) [inline code/core/src/components/components/Select/Select.tsx]: I know this was already here originally, but I think it would be better to extract this function.

**Review by ghengeveld — APPROVED** [review]
Not sure why CI didn't run. Code looks good, but I'd like to see how the Chromatic snapshots turn out. @MichaelArestad should probably take this for a spin as well.

**Comment by Sidnioulz** [comment]
> Not sure why CI didn't run. Code looks good, but I'd like to see how the Chromatic snapshots turn out. @MichaelArestad should probably take this for a spin as well.

I was hoping you wouldn't review this one so quickly, it's based on the IconButton/Select branch and I hadn't finished fixing all the E2E and unit tests on that. It should be green soon enough :)

File: code/core/src/components/components/Select/SelectOption.tsx (hunk 1 above)
Comment (Sidnioulz) [reply code/core/src/components/components/Select/SelectOption.tsx:64]: This one is essentially the 6px listbox radius - the padding + the outline offset so that the outline shape perfectly matches the listbox radius.

I'm not sure if there are variables for all of the above as I'm seeing so many hardcoded values.

@MichaelArestad should I revert this optimisation so we don't add cryptic/hardcoded code to the UI?

Also re: the DT project, this use case is why I usually have outline offset design tokens even though they can't be used in Figma. It helps make calculations like this more robust.

File: code/core/src/components/components/Select/Select.tsx [hunk 3]
```diff
@@ -135,15 +189,21 @@ export const Select = forwardRef<HTMLButtonElement, SelectProps>(
...
+    const triggerRef = useObjectRef(ref);
 
     const id = useId();
     const listboxId = `${id}-listbox`;
     const listboxRef = useRef<HTMLUListElement>(null);
 
+    const otState = useOverlayTriggerState({
+      isOpen: isOpen && !disabled,
+      onOpenChange: setIsOpen,
+    });
```
[the remark this answers is not in the document]
Comment (Sidnioulz) [reply code/core/src/components/components/Select/Select.tsx:206]: It's a hook call. We don't call hooks in DOM trees because whoever adds conditionals or early returns next could miss the hook call.

File: code/core/src/components/components/Select/SelectOption.tsx (hunk 1 above)
Comment (MichaelArestad) [reply code/core/src/components/components/Select/SelectOption.tsx:64]: @Sidnioulz I think it's okay. This is the kind of stuff we will clean up once we start introducing design tokens and being more rigorous with implementation.
